x/programs/cmd/simulator/src/lib.rs

```diff
@@ -227,105 +228,95 @@ impl Client {
-        .stdin(Stdio::piped())
-        .stdout(Stdio::piped())
-        .spawn()?;
-
-    // write json to stdin
-    let input =
-        serde_json::to_string(plan).map_err(|e| format!("failed to serialize json: {e}"))?;
-    if let Some(ref mut stdin) = child.stdin {
-        stdin
-            .write_all(input.as_bytes())
-            .map_err(|e| format!("failed to write to stdin: {e}"))?;
+impl Default for Client {
+    fn default() -> Self {
+        Self::new()
     }
+}
 
-    child
-        .wait_with_output()
-        .map_err(|e| format!("failed to wait for child: {e}").into())
+struct SimulatorChild {
+    stdin: ChildStdin,
+    lines: Lines<BufReader<ChildStdout>>,
 }
 
-fn run_steps<P, T>(path: P, plan: &Plan) -> Result<Vec<T>, Box<dyn Error>>
-where
-    T: serde::de::DeserializeOwned + serde::Serialize,
-    P: AsRef<OsStr>,
-{
-    let output = cmd_output(path, plan)?;
-    let mut items: Vec<T> = Vec::new();
-
-    if !output.status.success() {
-        println!("stderr");
-        for line in String::from_utf8_lossy(&output.stderr).lines() {
-            println!("{line}");
-        }
-        println!("stdout");
-        for line in String::from_utf8_lossy(&output.stdout).lines() {
-            println!("{line}");
-        }
-        return Err(String::from_utf8(output.stdout)?.into());
+impl SimulatorChild {
+    fn new(path: &str) -> Result<Self, Box<dyn Error>> {
+        let child = Command::new(path)
+            .arg("interpreter")
+            .arg("--cleanup")
+            .arg("--log-level")
+            .arg("error")
+            .stdin(Stdio::piped())
+            .stdout(Stdio::piped())
+            .spawn()?;
+
+        let (Some(stdin), Some(stdout)) = (child.stdin, child.stdout) else {
+            panic!("could not attach to the child process stdio");
+        };
+        let lines = BufReader::new(stdout).lines();
+
+        let mut sim = Self { stdin, lines };
+
+        let _ = sim.read_response()?; // intepreter ready
+
+        Ok(sim)
     }
 
-    for line in String::from_utf8(output.stdout)?
-        .lines()
-        .filter(|line| !line.trim().is_empty())
-    {
-        let item = serde_json::from_str(line)
-            .map_err(|e| format!("failed to parse output to json: {e}"))?;
-        items.push(item);
+    fn run_step(
+        &mut self,
+        caller_key: &String,
+        step: &Step,
+    ) -> Result<PlanResponse, Box<dyn Error>> {
+        let run_command = "run --stdin\n";
+        self.write_stdin(run_command.as_bytes())?;
+        let _ = self.read_response()?;
+
+        let step = SimulatorStep { caller_key, step };
+        let mut input =
+            serde_json::to_string(&step).map_err(|e| format!("failed to serialize json: {e}"))?;
+        input.push('\n');
+        self.write_stdin(input.as_bytes())?;
+        let resp = self.read_response()?;
+
+        Ok(resp)
     }
 
-    Ok(items)
-}
+    fn read_response(&mut self) -> Result<PlanResponse, Box<dyn Error>> {
```

Comment:
Instead of calling `format!` inline for errors, we should really just use `thiserror`.

Reply:
Done, `eyre` is great too because it gives meaningful backtraces when it crashes.